torch/utils/data/sampler.py

```diff
@@ -194,8 +194,8 @@ def __iter__(self) -> Iterator[int]:
             ).tolist()
         else:
             for _ in range(self.num_samples // n):
-                yield from torch.randperm(n, generator=generator).tolist()
-            yield from torch.randperm(n, generator=generator).tolist()[
+                yield from torch.randperm(n, generator=generator, device="cpu").tolist()
+            yield from torch.randperm(n, generator=generator, device="cpu").tolist()[
```

Comment:
Will this actually fix anything or would the generator have to be moved to the CPU?

Reply:
It definitely fixes issues when one runs:
```
with torch.device("cuda"):
    ...
    dl_train = DataLoader(dataset_train, shuffle=True, batch_size=batch_size)
    ...
```
because you can't run `.tolist()` on gpu tensors. Maybe, the best thing to do is to add `yield from torch.randperm(n, generator=generator).cpu().tolist()`, you tell me.

Reply:
I was mistaken about `.tolist()`. It does work on cuda tensors. Still, `device="cpu"` is the only fix I can provide.